src/common/column_matrix.h

```diff
@@ -13,310 +13,491 @@
 
     // store least bin id for each feature
-    index_base_ = const_cast<uint32_t*>(gmat.cut.Ptrs().data());
+    cut_ = gmat.cut;
+    index_base_ = const_cast<uint32_t*>(cut_.Ptrs().data());
 
-    any_missing_ = !gmat.IsDense();
+    const bool no_missing_values = NoMissingValues(gmat.row_ptr[n_row], n_row, n_feature);
+    any_missing_ = !no_missing_values;
 
     missing_flags_.clear();
+    if (no_missing_values) {
+      missing_flags_.resize(feature_offsets_[n_feature], false);
+    } else {
+      missing_flags_.resize(feature_offsets_[n_feature], true);
+    }
 
     // pre-fill index_ for dense columns
-    BinTypeSize gmat_bin_size = gmat.index.GetBinTypeSize();
-    if (!any_missing_) {
-      missing_flags_.resize(feature_offsets_[nfeature], false);
-      // row index is compressed, we need to dispatch it.
-      DispatchBinType(gmat_bin_size, [&, nrow, nfeature, n_threads](auto t) {
-        using RowBinIdxT = decltype(t);
-        SetIndexNoMissing(gmat.index.data<RowBinIdxT>(), nrow, nfeature, n_threads);
-      });
+    if (all_dense) {
+      BinTypeSize gmat_bin_size = gmat.index.GetBinTypeSize();
+      if (gmat_bin_size == kUint8BinsTypeSize) {
+        SetIndexAllDense(page, gmat.index.data<uint8_t>(), gmat, n_row, n_feature,
+                         no_missing_values, n_thread);
+      } else if (gmat_bin_size == kUint16BinsTypeSize) {
+        SetIndexAllDense(page, gmat.index.data<uint16_t>(), gmat, n_row, n_feature,
+                         no_missing_values, n_thread);
+      } else {
+        CHECK_EQ(gmat_bin_size, kUint32BinsTypeSize);
+        SetIndexAllDense(page, gmat.index.data<uint32_t>(), gmat, n_row, n_feature,
+                         no_missing_values, n_thread);
+      }
+      /* For sparse DMatrix gmat.index.getBinTypeSize() returns always kUint32BinsTypeSize
+         but for ColumnMatrix we still have a chance to reduce the memory consumption */
     } else {
-      missing_flags_.resize(feature_offsets_[nfeature], true);
-      SetIndexMixedColumns(batch, gmat.index.data<uint32_t>(), gmat, nfeature, missing);
+      if (bin_type_size_ == kUint8BinsTypeSize) {
+        SetIndex<uint8_t>(page, gmat.index.data<uint32_t>(), gmat, n_feature);
+      } else if (bin_type_size_ == kUint16BinsTypeSize) {
+        SetIndex<uint16_t>(page, gmat.index.data<uint32_t>(), gmat, n_feature);
+      } else {
+        CHECK_EQ(bin_type_size_, kUint32BinsTypeSize);
+        SetIndex<uint32_t>(page, gmat.index.data<uint32_t>(), gmat, n_feature);
+      }
     }
+    FillColumnViewList(n_feature);
   }
 
-  // construct column matrix from GHistIndexMatrix
-  void Init(SparsePage const& page, const GHistIndexMatrix& gmat, double sparse_threshold,
-            int32_t n_threads) {
-    auto batch = data::SparsePageAdapterBatch{page.GetView()};
-    this->Init(batch, std::numeric_limits<float>::quiet_NaN(), gmat, sparse_threshold, n_threads);
-  }
+  const ColumnListType& GetColumnList() const { return column_list_; }
```

Comment:
Yes, in opt_partition_builder.h (from main PR) we need to access many columns many times -> therefore a reference to the whole vector is returned to avoid additional perf cost (increment, decrement) caused by returning shared_ptr for each column.